tket/src/WeightSubgrMono/Reducing/CloseVerticesFilter.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/Reducing/CloseVerticesFilter.hpp"
+
+#include <cmath>
+#include <sstream>
+
+#include "WeightSubgrMono/Searching/FixedData.hpp"
+#include "WeightSubgrMono/Searching/SearchNodeWrapper.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+// In experiments, this appeared to be the best value
+// (in combination with all other reductions).
+CloseVerticesFilter::CloseVerticesFilter() : m_num_levels(2) {}
+
+bool CloseVerticesFilter::reduce(
+    const FixedData& fixed_data, Assignments& assignments,
+    std::size_t number_of_assignments_previously_processed_in_this_node,
```

Comment:
Yes, now the reducer objects DO each have their own internal counter.

Some complication is inevitable, though, since we need reducers etc. to be able to stop and restart from different points, etc.; it would be correct, but much slower, if we just did everything in one big lump with duplicated calculations.